nova/virt/vmwareapi/driver.py

```diff
@@ -753,6 +760,290 @@ def _server_group_sync_loop(self, compute_host):
+                           network_info, disk_info, migrate_data):
+        """Prepare an instance for live migration (on the destination host)"""
+        if migrate_data.instance_already_migrated:
+            return migrate_data
+
+        return self._pre_live_migration(context, instance,
+            block_device_info, network_info, disk_info, migrate_data)
+
+    def _pre_live_migration(self, context, instance, block_device_info,
+                            network_info, disk_info, migrate_data):
+        result = self._vmops.place_vm(context, instance)
+
+        if hasattr(result, 'drsFault'):
+            LOG.error("Placement Error: %s", vim_util.serialize_object(
+                result.drsFault), instance=instance)
+
+        if (not hasattr(result, 'recommendations') or
+                not result.recommendations):
+            raise exception.MigrationError(
+                reason="PlaceVM did not give any recommendations")
+
+        rs = sorted([r for r in result.recommendations
+                        if r.reason == "xvmotionPlacement" and
+                        r.action],
+                    key=attrgetter("rating"))
+        if not rs:
+            raise exception.MigrationError(
+                reason="Did not get any xvmotionPlacement")
+
+        relocate_spec = rs[0].action[0].relocateSpec
+
+        # Should never happen, but if it does we rather want an error
+        # here, than sometime down the line
+        if not relocate_spec.host:
+            raise exception.MigrationError(
+                reason="No host with enough resources")
+
+        # Samere here: Should never happen
+        if not relocate_spec.datastore:
+            raise exception.MigrationError(
+                reason="No datastore with enough resources")
+
+        # relocate_defaults are serialised/deserialised on put/get
+        defaults = migrate_data.relocate_defaults
+        spec = vim_util.serialize_object(relocate_spec)
+        defaults["relocate_spec"] = spec
+        # Writing the values back
+        migrate_data.relocate_defaults = defaults
+
+        return migrate_data
+
+    def _get_checked_volumes(self, context, instance, required_values,
+                             exc=exception.MigrationError):
+        volumes = self._get_volume_mappings(context, instance)
+        for key, volume in six.iteritems(volumes):
+            for v in required_values:
+                if v not in volume:
+                    message = "Missing {} for volume {} (Device {})".format(
+                        v, volume.get("volume_id"), key
+                    )
+                    raise exc(message)
+        return volumes
+
+    def live_migration(self, context, instance, dest,
+                       post_method, recover_method, block_migration=False,
+                       migrate_data=None):
+        """Live migration of an instance to another host."""
+        if migrate_data.instance_already_migrated:
```

Comment:
`migrate_data` is optional (`None` by default). This will fail if it's not set.

Reply:
Okay, I'll raise a `ValueError`, and log the error.